Map raw field names to attributes.

sankethkatta opened the pull request:
Add feature #3714. 

This also allows the `returning: true` [option](https://github.com/sequelize/sequelize/blob/2c4a9f3cf9887fb33c31e397e758dd4aa3374d01/lib/model.js#L2064) for `update` in postgres to work properly. Currently it [generates](https://github.com/sequelize/sequelize/blob/e948dc8fdc6be90c54441e091472a0c5d72899ce/lib/dialects/abstract/query-generator.js#L212) out `RETURNING *`, which faces the same problem as described with raw queries in #3714.

mickhansen commented:
I think it would be better if it happened at a lower level than build so it could be reused for bulkCreate etc.

sankethkatta commented:
Any suggestions where a better place for it would be? Maybe in the Instance constructor or initValues?

Also, not sure if you have any insight into the error the build system is throwing, I am not able to reproduce the error locally (and it seems unrelated to my changes): 
```
 Error: Expected 0 running queries. 1 queries still running in [POSTGRES-NATIVE] Continuation local storage bluebird shims then rejected
```

mickhansen commented:
Probably in post query result parsing, let the caller pass a field to attribute map that the query result parser can use perhaps. It's something we need a generalized approach to.

sankethkatta commented:
I attempted to generalize this a bit more and moved the aliasing into `handleSelectQuery`. It uses an option called `aliasMap` which is passed into `Sequelize.prototype.query`.
Let me know what you think.

mickhansen commented on lib/dialects/abstract/query.js:
```diff
@@ -514,7 +514,19 @@ AbstractQuery.prototype.isUpdateQuery = function() {
+      return Utils._.reduce(aliasMap, function(result, alias, field) {
+        if (field in result) {
+          result[alias] = result[field];
+          delete result[field];
```
`result[field] = undefined` is a bit quicker, can you check if it produces the same result?

sankethkatta replied:
It still passes the tests, however, it leaves remnants in the `dataValues` object.

```js
dataValues: { 
  id: 1,
  username: 'john',
  email_address: undefined,
  createdAt: Sun Jan 01 2012 02:10:10 GMT-0800 (PST),
  updatedAt: Sun Jan 01 2012 02:10:10 GMT-0800 (PST),
  emailAddress: 'john@gmail.com' 
}
```

mickhansen replied:
Ah yeah of course, we don't want that.

mickhansen commented on lib/dialects/abstract/query.js:
```diff
@@ -514,7 +514,19 @@ AbstractQuery.prototype.isUpdateQuery = function() {
+    var aliasMap = this.options.aliasMap;
+    results = Utils._.map(results, function(result) {
+      return Utils._.reduce(aliasMap, function(result, alias, field) {
+        if (field in result) {
```
does this syntax deopt @BridgeAR?

sankethkatta replied:
I can also use lodash's `_.has()` or `result.hasOwnProperty()`. I used the `in` syntax as I had seen in other places in the Sequelize codebase.

mickhansen commented:
Looking good. Although i'm not sure about the `aliasMap` name.

sankethkatta commented:
Looks like the other parameter names for `.query()` are single words like `replacements`. How about `aliases` or `mappings`?

mickhansen commented:
That's even less descriptive though :)

sankethkatta commented:
I see, more descriptive! How about `rawFieldMap`?

mickhansen commented:
Maybe just `fieldMap`? I think that would be okay.

sankethkatta commented:
That sounds good to me, updating it now.

BridgeAR commented on lib/dialects/abstract/query.js:
```diff
@@ -514,7 +514,19 @@ AbstractQuery.prototype.isUpdateQuery = function() {
+    var fieldMap = this.options.fieldMap;
+    results = Utils._.map(results, function(result) {
+      return Utils._.reduce(fieldMap, function(result, name, field) {
+        if (field in result) {
```
This should be written as 
```js
result.field !== undefined
```

mickhansen replied:
@BridgeAR `in` causes de-opt?

BridgeAR replied:
This depends on if any field name in result is somehow special or not. If it is e.g. `Strange / field` it'll deopt (there was an issue with someone using slashing in his field names :D ). Otherwise it should not but checking for undefined will be faster.

BridgeAR commented:
LGTM even tough I think it would be nice to improve this at some point to only include the fields needed from the beginning on. But this is only internal stuff.

mickhansen commented:
@BridgeAR From the beggining? I'm not quite sure i understand? You mean parse and rewrite a raw query to alias the fields?

BridgeAR commented:
Yes

sankethkatta commented:
Just pushed the minor change, checking if the field existed in the result using `undefined` as suggested.

mickhansen commented:
@BridgeAR That does not sound feasible at all, don't think we'll ever do that ;p Rewriting raw queries would be a bit too magic i think. (Why is the input not my actual input?)

mickhansen commented on lib/dialects/abstract/query.js:
```diff
@@ -514,7 +514,19 @@ AbstractQuery.prototype.isUpdateQuery = function() {
   var result = null;
-
+  // Map raw fields to arbitrary names if a mapping is provided
+  if (this.options.fieldMap) {
```
We should probably put this under the `raw` if so it isn't run on all selects, no need to encur the extra loop when it's not needed.

sankethkatta replied:
Currently the fix also works for postgres `returning: true` on `update`. We'd have to put it behind  `this.options.raw || this.options.type === QueryTypes.BULKUPDATE`.

mickhansen replied:
That sounds like a bug, `update` should not result in a handleSelectQuery call.

sankethkatta replied:
I assumed it did this by design as handling a `UPDATE ... RETURNING *` is similar to a select. Here is the call to `handleSelectQuery` in [postgres/query.js](https://github.com/sequelize/sequelize/blob/eaf197fa7e6cb4d6b52d3d13493e6ae81352fdf8/lib/dialects/postgres/query.js#L296).

mickhansen replied:
Huh, yeah maybe you're right. I'm afraid we can't incur that perf overhead for all select queries since they are already perf sensitive and wouldnt take nicely to another nested loop, please do the large if :)

sankethkatta replied:
Hmm, it looks like `this.options.raw` isn't set if a model is passed in [query.js#L671](https://github.com/sequelize/sequelize/blob/master/lib/sequelize.js#L671). Is there a way to know if the query originated from user input on `sequelize.query` directly without using `options.raw`?

mickhansen replied:
Hmm no, no way to know, that is a bit of a problem since actually in this case you don't want a raw result.

sankethkatta replied:
We could introduce a new option to toggle on or off the auto-mapping to model attributes behavior. Right now we assume we should map fields anytime the model option is set. 
But it would require users writing raw queries to set on the new option in addition to setting the model.

sankethkatta commented:
@mickhansen Trying another approach to this problem. I replaced `fieldMap` with a boolean `mapToModel` option. The mapping will no longer be on the code path for select queries, but will require users of the raw query api to explicitly set `mapToModel: true` to enable the behavior.

Let me know your thoughts!

janmeier commented:
@sankethkatta I like this approach, it looks pretty clean to me. I'll let @mickhansen comment on this too before merging but LGTM :)

sankethkatta commented:
@janmeier @mickhansen any update on this?

janmeier commented:
@sankethkatta Mick is on vacation, I'll bug him about this when he's back ;)

sankethkatta commented:
Thanks! :smiley:

sankethkatta commented:
ping! :bell:

mattppp commented:
Merge please?

mickhansen commented:
Sorry i was on a trip again.
I'd prefer if we do as little as coupling as possible in the query generator, so ideally generate the field map in `query` that can be used by `handleSelectQuery` perhaps.

mickhansen commented:
Hmm, nvm the last comment i just deleted.

sankethkatta commented:
I don't like that the current approach requires the user to set  the `mapToModel` option. However, given the performance limitations of running unnecessary loops and the fact it is not possible to know whether `sequelize.query` was called directly by the user (which we discussed [pull/3995#discussion_r34085726] (https://github.com/sequelize/sequelize/pull/3995#discussion_r34085726)). This seems like the best compromise, and I'd rather err on the side of explicitness anyways.

Let me know if you can think of a better way!

mickhansen commented on lib/dialects/abstract/query.js:
```diff
@@ -514,7 +514,19 @@ AbstractQuery.prototype.isUpdateQuery = function() {
+        return result;
+      }, result);
+    });
+  }
```
Sorry for being so late to comment.
Ideally i'd like a generic mapper passed to query, or one generated in query and then used here.

The less the query generator and query handler needs to know about `Model` the better.

sankethkatta replied:
Refactored in latest commit.

mickhansen commented on changelog.md:
```diff
@@ -4,6 +4,7 @@
 - [FIXED] Reload doesn't synchronize a null include [#4353](https://github.com/sequelize/sequelize/issues/4353)
 - [FIXED] commit/rollback multiple times on same transaction [#4491](https://github.com/sequelize/sequelize/issues/4491)
 - [FIXED] memory leak / options mangle for scopes with include [#4470](https://github.com/sequelize/sequelize/issues/4470)
+- [FEATURE] Map raw fields back to attributes when model is built [#3714](https://github.com/sequelize/sequelize/issues/3714)
```
`[FEATURE]` > `[ADDED]`
perhaps: "Map raw fields back to attribute names when using `returning` ..."

mickhansen commented on lib/sequelize.js:
```diff
@@ -659,6 +661,11 @@ Sequelize.prototype.query = function(sql, options) {
   }
 
+  // Map raw fields to model field names using the `fieldAttributeMap`
+  if (options.model && options.mapToModel && Utils._.size(options.model.fieldAttributeMap) > 0) {
```
Nitpicking, but could use `!_.empty(..)` i believe?

sankethkatta replied:
Good call!

mickhansen commented:
Please add API docs to `sequelize.query` describing `fieldMap` and `mapToModel` options.
Excellent work!

(please squash commits to a single descriptive commit when ready aswell)

sankethkatta commented:
I addressed the comments and squashed the commits. 
As for the documentation, do I need to write somewhere else in addition to the JSDoc? (which I did on [`lib/sequelize#L643`](https://github.com/sequelize/sequelize/pull/3995/files#diff-24f8b5cb05b7e92726ca732f7d16d5baR643).
Thanks!

mickhansen commented:
Nope jsdoc works :)

mickhansen commented:
Thank you for your work!

sankethkatta commented:
Awesome! Thank you for working through this with me! :smiley:
